enhancements/network/ip-interface-selection.md

```diff
@@ -0,0 +1,285 @@
+  Can we just require them to use this mechanism if they want custom IP
+  selection?
+
+- The interface file written by nodeip-configuration will no longer be valid
+  once configure-ovs moves the IP to br-ex. Should we re-run
+  nodeip-configuration at that point to update it, should be leave the
+  interface file alone, or should we just not write the interface at all and
+  use the IP address(es) exclusively for interface selection?
+  - Answer: Just don't persist the interface name. As long the IP used to get
+    the interface is consistent between services we should get the behavior
+    we need. Also, this option is more flexible for future features, like the
+    ability to run dual stack clusters with ipv4 and ipv6 on different
+    interfaces.
+
+- We may want to rename KUBELET_NODEIP_HINT to reflect the fact that it will
+  now affect more than just Kubelet. Is just NODEIP_HINT acceptable? How do we
+  handle backward compatibility with the KUBELET version of the name to avoid
+  breaking existing users of that functionality? I'm not sure how much logic
+  we can inject in the systemd service.
+
+### Test Plan
+
+In general this will be covered by existing e2e tests. However, we should
+expand coverage of the nodeip-configuration and configure-ovs components as
+they have proven to have a lot of edge cases.
+
+At this point I'm not sure exactly what a more-targeted set of tests for
+host networking would look like though. We likely cannot stand up the large
+variety of networking environments needed to run e2e tests with all of the
+possible architectures, so we need something more like a unit or functional
+test.
+
+Given that this work will not make the testing situation any worse, it's
+possible we could defer the testing improvements to followup work.
+
+### Graduation Criteria
+
+NA
+
+#### Dev Preview -> Tech Preview
+
+NA
+
+#### Tech Preview -> GA
+
+NA
+
+#### Removing a deprecated feature
+
+NA
+
+### Upgrade / Downgrade Strategy
+
+NA
+
+### Version Skew Strategy
+
+From version to version the selection process must remain consistent in order
+to avoid IPs and interfaces changing. As a result, version skew should not be
+a problem.
+
+### Operational Aspects of API Extensions
+
+NA
+
+#### Failure Modes
+
+NA
+
+#### Support Procedures
+
+This should not drastically affect support. It will still be possible to
+determine what address/interface a service is using by looking at its config
+and/or logs. The only change will be where that comes from and that the new
+files can also be consulted to check the behavior of nodeip-configuration.
+
+## Implementation History
+
+KUBELET_NODEIP_HINT was implemented in 4.11 and backported to older releases
+to improve the user experience where they needed to override the default logic
+in nodeip-configuration.
+
+## Alternatives
+
+Leave things basically as they are, but teach services like configure-ovs and
```

Comment:
Right now user should be able to configure the network how they want, although rather undocumented. It involves:

* Deploying their own keyfiles configuring br-ex on their own and using different filenames than what configure-ovs would use
* In 4.10 release or later, disabling configure-ovs (the service itself, or touching /run/configure-ovs-boot-done)

I guess one alternative would be to refine & document this.

Reply:
Yeah, I guess that in combination with KUBELET_NODEIP_HINT might handle a number of the use cases, but I feel like manual br-ex creation is a big hammer for a relatively small nail. I can add that to the alternatives though in case we decide to pursue that instead.